File: paimon-common/src/main/java/org/apache/paimon/globalindex/SortedFileMetaSelector.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.globalindex;

import org.apache.paimon.memory.MemorySlice;
import org.apache.paimon.predicate.FieldRef;
import org.apache.paimon.predicate.FunctionVisitor;
import org.apache.paimon.predicate.LeafPredicate;
import org.apache.paimon.utils.Pair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * Selects candidate global index files by per-index-file min/max metadata.
 *
 * <p>All files are expected to belong to the same field. The current {@code
 * RowRangeGlobalIndexScanner} can guarantee that. Please do not break this premise if you want to
 * implement your own index scanner.
 */
public class SortedFileMetaSelector implements FunctionVisitor<Optional<List<GlobalIndexIOMeta>>> {

    private final List<Pair<GlobalIndexIOMeta, SortedIndexFileMeta>> files;
    private final KeySerializer keySerializer;
    private final Comparator<Object> comparator;

    public SortedFileMetaSelector(List<GlobalIndexIOMeta> files, KeySerializer keySerializer) {
        this.files =
                files.stream()
                        .map(
                                meta ->
                                        Pair.of(
                                                meta,
                                                SortedIndexFileMeta.deserialize(meta.metadata())))
                        .collect(Collectors.toList());
        this.keySerializer = keySerializer;
        this.comparator = keySerializer.createComparator();
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitIsNotNull(FieldRef fieldRef) {
        return Optional.of(filter(meta -> !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitIsNull(FieldRef fieldRef) {
        return Optional.of(filter(SortedIndexFileMeta::hasNulls));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitStartsWith(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }

        byte[] prefix = serialize(literal);
        if (prefix.length == 0) {
            return Optional.of(filter(meta -> !meta.onlyNulls()));
        }

        Object prefixKey = deserialize(prefix);
        byte[] upperBoundBytes = prefixUpperBound(prefix);
        Object upperBound = upperBoundBytes == null ? null : deserialize(upperBoundBytes);
        return Optional.of(
                filter(
                        meta ->
                                !meta.onlyNulls()
                                        && compareLastKey(meta, prefixKey) >= 0
                                        && (upperBound == null
                                                || compareFirstKey(meta, upperBound) < 0)));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitEndsWith(FieldRef fieldRef, Object literal) {
        return Optional.of(filter(meta -> literal != null && !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitContains(FieldRef fieldRef, Object literal) {
        return Optional.of(filter(meta -> literal != null && !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitLike(FieldRef fieldRef, Object literal) {
        return Optional.of(filter(meta -> literal != null && !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitLessThan(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls() && compareFirstKey(meta, literal) < 0));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitGreaterOrEqual(
            FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls() && compareLastKey(meta, literal) >= 0));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitNotEqual(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitLessOrEqual(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(
                filter(meta -> !meta.onlyNulls() && compareFirstKey(meta, literal) <= 0));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitEqual(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls() && overlaps(meta, literal, literal)));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitGreaterThan(FieldRef fieldRef, Object literal) {
        if (literal == null) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls() && compareLastKey(meta, literal) > 0));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitIn(FieldRef fieldRef, List<Object> literals) {
        return Optional.of(
                filter(
                        meta -> {
                            if (meta.onlyNulls()) {
                                return false;
                            }
                            for (Object literal : literals) {
                                if (literal != null && overlaps(meta, literal, literal)) {
                                    return true;
                                }
                            }
                            return false;
                        }));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitNotIn(FieldRef fieldRef, List<Object> literals) {
        for (Object literal : literals) {
            if (literal == null) {
                return Optional.of(Collections.emptyList());
            }
        }
        return Optional.of(filter(meta -> !meta.onlyNulls()));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitBetween(
            FieldRef fieldRef, Object from, Object to) {
        if (from == null || to == null || compareKeys(from, to) > 0) {
            return Optional.of(Collections.emptyList());
        }
        return Optional.of(filter(meta -> !meta.onlyNulls() && overlaps(meta, from, to)));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitAnd(
            List<Optional<List<GlobalIndexIOMeta>>> children) {
        HashSet<GlobalIndexIOMeta> result = null;
        for (Optional<List<GlobalIndexIOMeta>> child : children) {
            if (!child.isPresent()) {
                return Optional.empty();
            }
            if (result == null) {
                result = new HashSet<>(child.get());
            } else {
                result.retainAll(child.get());
            }
            if (result.isEmpty()) {
                break;
            }
        }
        return result == null ? Optional.empty() : Optional.of(new ArrayList<>(result));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitOr(
            List<Optional<List<GlobalIndexIOMeta>>> children) {
        HashSet<GlobalIndexIOMeta> result = new HashSet<>();
        for (Optional<List<GlobalIndexIOMeta>> child : children) {
            if (!child.isPresent()) {
                return Optional.empty();
            }
            result.addAll(child.get());
        }
        return Optional.of(new ArrayList<>(result));
    }

    @Override
    public Optional<List<GlobalIndexIOMeta>> visitNonFieldLeaf(LeafPredicate predicate) {
        return Optional.empty();
    }

    protected boolean overlaps(SortedIndexFileMeta meta, Object from, Object to) {
        return comparator.compare(from, deserialize(meta.lastKey())) <= 0
                && comparator.compare(to, deserialize(meta.firstKey())) >= 0;
    }

    protected int compareFirstKey(SortedIndexFileMeta meta, Object literal) {
        return comparator.compare(deserialize(meta.firstKey()), literal);
    }

    protected int compareLastKey(SortedIndexFileMeta meta, Object literal) {
        return comparator.compare(deserialize(meta.lastKey()), literal);
    }

    protected int compareKeys(Object left, Object right) {
        return comparator.compare(left, right);
    }

    protected byte[] serialize(Object key) {
        return keySerializer.serialize(key);
    }

    protected Object deserialize(byte[] valueBytes) {
        return keySerializer.deserialize(MemorySlice.wrap(valueBytes));
    }

    protected List<GlobalIndexIOMeta> filter(Predicate<SortedIndexFileMeta> predicate) {
        return files.stream()
                .filter(pair -> predicate.test(pair.getRight()))
                .map(Pair::getLeft)
                .collect(Collectors.toList());
    }

    protected static byte[] prefixUpperBound(byte[] prefix) {
        for (int i = prefix.length - 1; i >= 0; i--) {
            int unsignedByte = prefix[i] & 0xFF;
            if (unsignedByte != 0xFF) {
                byte[] upperBound = new byte[i + 1];
                System.arraycopy(prefix, 0, upperBound, 0, i + 1);
                upperBound[i] = (byte) (unsignedByte + 1);
                return upperBound;
            }
        }
        return null;
    }
}

File: paimon-python/pypaimon/globalindex/sorted_file_meta_selector.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Prunes sorted global index files by manifest min/max metadata."""

from typing import List, Optional, Tuple

from pypaimon.globalindex.global_index_meta import GlobalIndexIOMeta
from pypaimon.globalindex.key_serializer import KeySerializer
from pypaimon.globalindex.sorted_index_file_meta import SortedIndexFileMeta


class SortedFileMetaSelector:
    """Selects candidate sorted global index files for a predicate."""

    def __init__(
        self,
        files: List[Tuple[GlobalIndexIOMeta, SortedIndexFileMeta]],
        key_serializer: KeySerializer,
    ):
        self._files = files
        self._key_serializer = key_serializer
        self._comparator = key_serializer.create_comparator()

    def _filter(self, predicate) -> List[GlobalIndexIOMeta]:
        return [io_meta for io_meta, idx_meta in self._files if predicate(idx_meta)]

    def select_is_null(self) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: meta.has_nulls)

    def select_is_not_null(self) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: not meta.only_nulls())

    def select_starts_with(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []

        prefix = self._key_serializer.serialize(literal)
        if len(prefix) == 0:
            return self._filter(lambda meta: not meta.only_nulls())

        prefix_key = self._key_serializer.deserialize(prefix)
        upper_bound_bytes = self._prefix_upper_bound(prefix)
        upper_bound = (None if upper_bound_bytes is None
                       else self._key_serializer.deserialize(upper_bound_bytes))

        return self._filter(
            lambda meta: (
                not meta.only_nulls()
                and self._compare_last_key(meta, prefix_key) >= 0
                and (upper_bound is None
                     or self._compare_first_key(meta, upper_bound) < 0)))

    def select_ends_with(self, literal) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: literal is not None and not meta.only_nulls())

    def select_contains(self, literal) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: literal is not None and not meta.only_nulls())

    def select_like(self, literal) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: literal is not None and not meta.only_nulls())

    def select_less_than(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._compare_first_key(meta, literal) < 0)

    def select_greater_or_equal(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._compare_last_key(meta, literal) >= 0)

    def select_not_equal(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(lambda meta: not meta.only_nulls())

    def select_less_or_equal(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._compare_first_key(meta, literal) <= 0)

    def select_equal(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._overlaps(meta, literal, literal))

    def select_greater_than(self, literal) -> List[GlobalIndexIOMeta]:
        if literal is None:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._compare_last_key(meta, literal) > 0)

    def select_in(self, literals) -> List[GlobalIndexIOMeta]:
        return self._filter(lambda meta: self._matches_any_literal(meta, literals))

    def select_not_in(self, literals) -> List[GlobalIndexIOMeta]:
        if any(literal is None for literal in literals):
            return []
        return self._filter(lambda meta: not meta.only_nulls())

    def select_between(self, from_v, to_v) -> List[GlobalIndexIOMeta]:
        if from_v is None or to_v is None or self._comparator(from_v, to_v) > 0:
            return []
        return self._filter(
            lambda meta: not meta.only_nulls()
            and self._overlaps(meta, from_v, to_v))

    def select_or(self, children: List[Optional[List[GlobalIndexIOMeta]]]
                  ) -> Optional[List[GlobalIndexIOMeta]]:
        result = {}
        for child in children:
            if child is None:
                return None
            for io_meta in child:
                key = io_meta.external_path or io_meta.file_name
                result[key] = io_meta
        return list(result.values())

    def _matches_any_literal(self, meta: SortedIndexFileMeta, literals) -> bool:
        if meta.only_nulls():
            return False
        for literal in literals:
            if literal is not None and self._overlaps(meta, literal, literal):
                return True
        return False

    def _overlaps(self, meta: SortedIndexFileMeta, from_v, to_v) -> bool:
        return (self._comparator(from_v, self._key_serializer.deserialize(meta.last_key)) <= 0
                and self._comparator(to_v, self._key_serializer.deserialize(meta.first_key)) >= 0)

    def _compare_first_key(self, meta: SortedIndexFileMeta, literal) -> int:
        return self._comparator(self._key_serializer.deserialize(meta.first_key), literal)

    def _compare_last_key(self, meta: SortedIndexFileMeta, literal) -> int:
        return self._comparator(self._key_serializer.deserialize(meta.last_key), literal)

    @staticmethod
    def _prefix_upper_bound(prefix: bytes) -> Optional[bytes]:
        for index in range(len(prefix) - 1, -1, -1):
            unsigned_byte = prefix[index]
            if unsigned_byte != 0xFF:
                upper_bound = bytearray(prefix[:index + 1])
                upper_bound[index] = unsigned_byte + 1
                return bytes(upper_bound)
        return None
